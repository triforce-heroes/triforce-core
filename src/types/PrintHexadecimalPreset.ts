export const enum PrintHexadecimalPreset {
  SIMPLIFIED,
  UINT8,
  UINT16_LE,
  UINT16_BE,
  UINT32_LE,
  UINT32_BE,
  INT8,
  INT16_LE,
  INT16_BE,
  INT32_LE,
  INT32_BE,
  FLOAT_LE,
  FLOAT_BE,
}
