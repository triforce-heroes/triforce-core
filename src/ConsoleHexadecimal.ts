#!/usr/bin/env node

import chalk from "chalk";

import { PrintHexadecimalPreset } from "@/types/PrintHexadecimalPreset.js";

const hexadecimalDot = chalk.ansi256(240)(".");
const hexadecimalNumberPadding = " ".repeat(9);

type PrintHexadecimalPresetData = [
  padding: number,
  bytes: number,
  unsigned?: boolean,
  littleEndian?: boolean,
  isFloat?: boolean,
];

const printHexadecimalPreset: Record<
  PrintHexadecimalPreset,
  PrintHexadecimalPresetData
> = {
  [PrintHexadecimalPreset.SIMPLIFIED]: [2, 1],
  [PrintHexadecimalPreset.UINT8]: [4, 1, true],
  [PrintHexadecimalPreset.UINT16_LE]: [3, 2, true, true],
  [PrintHexadecimalPreset.UINT16_BE]: [3, 2, true],
  [PrintHexadecimalPreset.UINT32_LE]: [3, 4, true, true],
  [PrintHexadecimalPreset.UINT32_BE]: [3, 4, true],
  [PrintHexadecimalPreset.INT8]: [5, 1],
  [PrintHexadecimalPreset.INT16_LE]: [4, 2, false, true],
  [PrintHexadecimalPreset.INT16_BE]: [4, 2],
  [PrintHexadecimalPreset.INT32_LE]: [3, 4, false, true],
  [PrintHexadecimalPreset.INT32_BE]: [3, 4],
  [PrintHexadecimalPreset.FLOAT_LE]: [7, 4, undefined, true, true],
  [PrintHexadecimalPreset.FLOAT_BE]: [7, 4, undefined, false, true],
};

function numberNormalize(number: number) {
  const numberString = String(number);

  if (!numberString.includes(".") || numberString.includes("e")) {
    return numberString;
  }

  return number.toFixed(15 - String(Math.floor(number)).length);
}

function generateNumbers(buffer: Buffer, preset: PrintHexadecimalPreset) {
  let numbers = hexadecimalNumberPadding;

  const [padding, bytes, unsigned, littleEndian, isFloat] =
    printHexadecimalPreset[preset];

  const bufferFunction =
    isFloat === true
      ? littleEndian === true
        ? "readFloatLE"
        : "readFloatBE"
      : littleEndian === true
      ? unsigned === true
        ? "readUIntLE"
        : "readIntLE"
      : unsigned === true
      ? "readUIntBE"
      : "readIntBE";

  for (let j = 0; j <= buffer.length - bytes; j += bytes) {
    numbers += numberNormalize(buffer[bufferFunction](j, bytes)).padStart(
      padding * bytes,
      " ",
    );
  }

  numbers += "\n";

  return chalk.gray(numbers);
}

export function printHexadecimal(
  buffer: Buffer,
  preset = PrintHexadecimalPreset.SIMPLIFIED,
) {
  const [padding] = printHexadecimalPreset[preset];
  let output = "";

  for (let i = 0; i < buffer.length; i += 16) {
    let bufferOutput = chalk.gray(
      `${i.toString(16).toUpperCase().padStart(8, "0")} `,
    );

    let hexadecimals = "";
    let characters = "";

    const bufferDigits = buffer.subarray(i, i + 16);

    for (const bufferDigit of bufferDigits) {
      hexadecimals += bufferDigit
        .toString(16)
        .toUpperCase()
        .padStart(2, "0")
        .padStart(padding, " ");

      characters +=
        (bufferDigit > 31 && bufferDigit < 127) ||
        (bufferDigit > 159 && bufferDigit <= 255)
          ? String.fromCodePoint(bufferDigit)
          : hexadecimalDot;
    }

    bufferOutput += hexadecimals.padEnd(padding * 16 + 1, " ");
    bufferOutput += chalk.whiteBright(characters);
    bufferOutput += "\n";

    output +=
      preset === PrintHexadecimalPreset.SIMPLIFIED
        ? bufferOutput
        : bufferOutput + generateNumbers(bufferDigits, preset);
  }

  process.stdout.write(output);
}
