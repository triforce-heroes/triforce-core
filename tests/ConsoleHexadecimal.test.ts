import chalk from "chalk";
import strip from "strip-ansi";
import { describe, expect, it, vitest } from "vitest";

import { printHexadecimal } from "../src/ConsoleHexadecimal.js";
import { PrintHexadecimalPreset } from "../src/types/PrintHexadecimalPreset.js";

describe("console", () => {
  chalk.level = 2;

  const samples = [
    ["simplified empty", Buffer.from([]), PrintHexadecimalPreset.SIMPLIFIED],
    ["simplified 1 byte", Buffer.from([0]), PrintHexadecimalPreset.SIMPLIFIED],
    [
      "simplified 40 bytes",
      Buffer.from("hello".repeat(8)),
      PrintHexadecimalPreset.SIMPLIFIED,
    ],
    ["UINT8 1 byte", Buffer.from([0]), PrintHexadecimalPreset.UINT8],
    ["INT8 1 byte", Buffer.from([0]), PrintHexadecimalPreset.INT8],
    [
      "UINT8 5 bytes",
      Buffer.from([10, 20, 30, 40, 250]),
      PrintHexadecimalPreset.UINT8,
    ],
    [
      "INT8 5 bytes",
      Buffer.from([10, 20, 30, 40, 250]),
      PrintHexadecimalPreset.INT8,
    ],
    [
      "INT8 26 bytes",
      Buffer.from([
        0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150,
        160, 170, 180, 190, 200, 210, 220, 230, 240, 250,
      ]),
      PrintHexadecimalPreset.INT8,
    ],
    [
      "UINT16_LE 5 bytes",
      Buffer.from([10, 20, 30, 250, 250]),
      PrintHexadecimalPreset.UINT16_LE,
    ],
    [
      "INT16_LE 5 bytes",
      Buffer.from([10, 20, 30, 250, 250]),
      PrintHexadecimalPreset.INT16_LE,
    ],
    [
      "UINT32_LE 5 bytes",
      Buffer.from([10, 20, 30, 250, 250]),
      PrintHexadecimalPreset.UINT32_LE,
    ],
    [
      "INT32_LE 5 bytes",
      Buffer.from([10, 20, 30, 250, 250]),
      PrintHexadecimalPreset.INT32_LE,
    ],
    [
      "FLOAT_LE 9 bytes",
      Buffer.from([0x42, 0xf6, 0xe9, 0x79, 0x79, 0xe9, 0xf6, 0x42, 0x00]),
      PrintHexadecimalPreset.FLOAT_LE,
    ],
    [
      "UINT16_BE 5 bytes",
      Buffer.from([10, 20, 30, 250, 250]),
      PrintHexadecimalPreset.UINT16_BE,
    ],
    [
      "INT16_BE 5 bytes",
      Buffer.from([10, 20, 30, 250, 250]),
      PrintHexadecimalPreset.INT16_BE,
    ],
    [
      "UINT32_BE 5 bytes",
      Buffer.from([10, 20, 30, 250, 250]),
      PrintHexadecimalPreset.UINT32_BE,
    ],
    [
      "INT32_BE 5 bytes",
      Buffer.from([10, 20, 30, 250, 250]),
      PrintHexadecimalPreset.INT32_BE,
    ],
    [
      "FLOAT_BE 9 bytes",
      Buffer.from([0x42, 0xf6, 0xe9, 0x79, 0x79, 0xe9, 0xf6, 0x42, 0x00]),
      PrintHexadecimalPreset.FLOAT_BE,
    ],
  ] as const;

  it.each(samples)("function printHexadecimal(): %s", (_, buffer, preset) => {
    let stdoutMessage: Uint8Array | string | undefined;

    vitest
      .spyOn(process.stdout, "write")
      .mockImplementationOnce((str: Uint8Array | string) => {
        stdoutMessage = str;

        return true;
      });

    printHexadecimal(buffer, preset);

    expect(strip(stdoutMessage as string)).toMatchSnapshot("ansi");
    expect(stdoutMessage).toMatchSnapshot("colors");
  });
});
