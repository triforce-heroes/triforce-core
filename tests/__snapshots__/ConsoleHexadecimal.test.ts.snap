// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`console > function printHexadecimal(): FLOAT_BE 9 bytes > ansi 1`] = `
"00000000      42     F6     E9     79     79     E9     F6     42     00                                                  BÃ¶Ã©yyÃ©Ã¶B.
                     123.456001281738      1.5184998373247989e+35
"
`;

exports[`console > function printHexadecimal(): FLOAT_BE 9 bytes > colors 1`] = `
"[90m00000000 [39m     42     F6     E9     79     79     E9     F6     42     00                                                  [97mBÃ¶Ã©yyÃ©Ã¶B[38;5;240m.[39m[97m[39m
[90m                     123.456001281738      1.5184998373247989e+35[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): FLOAT_LE 9 bytes > ansi 1`] = `
"00000000      42     F6     E9     79     79     E9     F6     42     00                                                  BÃ¶Ã©yyÃ©Ã¶B.
               1.5184998373247989e+35            123.456001281738
"
`;

exports[`console > function printHexadecimal(): FLOAT_LE 9 bytes > colors 1`] = `
"[90m00000000 [39m     42     F6     E9     79     79     E9     F6     42     00                                                  [97mBÃ¶Ã©yyÃ©Ã¶B[38;5;240m.[39m[97m[39m
[90m               1.5184998373247989e+35            123.456001281738[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): INT8 1 byte > ansi 1`] = `
"00000000    00                                                                            .
             0
"
`;

exports[`console > function printHexadecimal(): INT8 1 byte > colors 1`] = `
"[90m00000000 [39m   00                                                                            [97m[38;5;240m.[39m[97m[39m
[90m             0[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): INT8 5 bytes > ansi 1`] = `
"00000000    0A   14   1E   28   FA                                                        ...(Ãº
            10   20   30   40   -6
"
`;

exports[`console > function printHexadecimal(): INT8 5 bytes > colors 1`] = `
"[90m00000000 [39m   0A   14   1E   28   FA                                                        [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m(Ãº[39m
[90m            10   20   30   40   -6[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): INT8 26 bytes > ansi 1`] = `
"00000000    00   0A   14   1E   28   32   3C   46   50   5A   64   6E   78   82   8C   96 ....(2<FPZdnx...
             0   10   20   30   40   50   60   70   80   90  100  110  120 -126 -116 -106
00000010    A0   AA   B4   BE   C8   D2   DC   E6   F0   FA                               Â ÂªÂ´Â¾ÃˆÃ’ÃœÃ¦Ã°Ãº
           -96  -86  -76  -66  -56  -46  -36  -26  -16   -6
"
`;

exports[`console > function printHexadecimal(): INT8 26 bytes > colors 1`] = `
"[90m00000000 [39m   00   0A   14   1E   28   32   3C   46   50   5A   64   6E   78   82   8C   96 [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m(2<FPZdnx[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[39m
[90m             0   10   20   30   40   50   60   70   80   90  100  110  120 -126 -116 -106[39m
[90m[39m[90m00000010 [39m   A0   AA   B4   BE   C8   D2   DC   E6   F0   FA                               [97mÂ ÂªÂ´Â¾ÃˆÃ’ÃœÃ¦Ã°Ãº[39m
[90m           -96  -86  -76  -66  -56  -46  -36  -26  -16   -6[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): INT16_BE 5 bytes > ansi 1`] = `
"00000000   0A  14  1E  FA  FA                                             ...ÃºÃº
             2580    7930
"
`;

exports[`console > function printHexadecimal(): INT16_BE 5 bytes > colors 1`] = `
"[90m00000000 [39m  0A  14  1E  FA  FA                                             [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97mÃºÃº[39m
[90m             2580    7930[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): INT16_LE 5 bytes > ansi 1`] = `
"00000000   0A  14  1E  FA  FA                                             ...ÃºÃº
             5130   -1506
"
`;

exports[`console > function printHexadecimal(): INT16_LE 5 bytes > colors 1`] = `
"[90m00000000 [39m  0A  14  1E  FA  FA                                             [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97mÃºÃº[39m
[90m             5130   -1506[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): INT32_BE 5 bytes > ansi 1`] = `
"00000000  0A 14 1E FA FA                                  ...ÃºÃº
            169090810
"
`;

exports[`console > function printHexadecimal(): INT32_BE 5 bytes > colors 1`] = `
"[90m00000000 [39m 0A 14 1E FA FA                                  [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97mÃºÃº[39m
[90m            169090810[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): INT32_LE 5 bytes > ansi 1`] = `
"00000000  0A 14 1E FA FA                                  ...ÃºÃº
            -98692086
"
`;

exports[`console > function printHexadecimal(): INT32_LE 5 bytes > colors 1`] = `
"[90m00000000 [39m 0A 14 1E FA FA                                  [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97mÃºÃº[39m
[90m            -98692086[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): UINT8 1 byte > ansi 1`] = `
"00000000   00                                                             .
            0
"
`;

exports[`console > function printHexadecimal(): UINT8 1 byte > colors 1`] = `
"[90m00000000 [39m  00                                                             [97m[38;5;240m.[39m[97m[39m
[90m            0[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): UINT8 5 bytes > ansi 1`] = `
"00000000   0A  14  1E  28  FA                                             ...(Ãº
           10  20  30  40 250
"
`;

exports[`console > function printHexadecimal(): UINT8 5 bytes > colors 1`] = `
"[90m00000000 [39m  0A  14  1E  28  FA                                             [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m(Ãº[39m
[90m           10  20  30  40 250[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): UINT16_BE 5 bytes > ansi 1`] = `
"00000000  0A 14 1E FA FA                                  ...ÃºÃº
           2580  7930
"
`;

exports[`console > function printHexadecimal(): UINT16_BE 5 bytes > colors 1`] = `
"[90m00000000 [39m 0A 14 1E FA FA                                  [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97mÃºÃº[39m
[90m           2580  7930[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): UINT16_LE 5 bytes > ansi 1`] = `
"00000000  0A 14 1E FA FA                                  ...ÃºÃº
           5130 64030
"
`;

exports[`console > function printHexadecimal(): UINT16_LE 5 bytes > colors 1`] = `
"[90m00000000 [39m 0A 14 1E FA FA                                  [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97mÃºÃº[39m
[90m           5130 64030[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): UINT32_BE 5 bytes > ansi 1`] = `
"00000000  0A 14 1E FA FA                                  ...ÃºÃº
            169090810
"
`;

exports[`console > function printHexadecimal(): UINT32_BE 5 bytes > colors 1`] = `
"[90m00000000 [39m 0A 14 1E FA FA                                  [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97mÃºÃº[39m
[90m            169090810[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): UINT32_LE 5 bytes > ansi 1`] = `
"00000000  0A 14 1E FA FA                                  ...ÃºÃº
           4196275210
"
`;

exports[`console > function printHexadecimal(): UINT32_LE 5 bytes > colors 1`] = `
"[90m00000000 [39m 0A 14 1E FA FA                                  [97m[38;5;240m.[39m[97m[38;5;240m.[39m[97m[38;5;240m.[39m[97mÃºÃº[39m
[90m           4196275210[39m
[90m[39m"
`;

exports[`console > function printHexadecimal(): simplified 1 byte > ansi 1`] = `
"00000000 00                               .
"
`;

exports[`console > function printHexadecimal(): simplified 1 byte > colors 1`] = `
"[90m00000000 [39m00                               [97m[38;5;240m.[39m[97m[39m
"
`;

exports[`console > function printHexadecimal(): simplified 40 bytes > ansi 1`] = `
"00000000 68656C6C6F68656C6C6F68656C6C6F68 hellohellohelloh
00000010 656C6C6F68656C6C6F68656C6C6F6865 ellohellohellohe
00000020 6C6C6F68656C6C6F                 llohello
"
`;

exports[`console > function printHexadecimal(): simplified 40 bytes > colors 1`] = `
"[90m00000000 [39m68656C6C6F68656C6C6F68656C6C6F68 [97mhellohellohelloh[39m
[90m00000010 [39m656C6C6F68656C6C6F68656C6C6F6865 [97mellohellohellohe[39m
[90m00000020 [39m6C6C6F68656C6C6F                 [97mllohello[39m
"
`;

exports[`console > function printHexadecimal(): simplified empty > ansi 1`] = `""`;

exports[`console > function printHexadecimal(): simplified empty > colors 1`] = `""`;
